Smoke Tests Codegen Working

sbera87 opened the pull request:
*Issue #, if available:*

*Description of changes:*

*Check all that applies:*
- [x] Did a review by yourself.
- [ ] Added proper tests to cover this PR. (If tests are not applicable, explain.)
- [ ] Checked if this PR is a breaking (APIs have been changed) change.
- [ ] Checked if this PR will _not_ introduce cross-platform inconsistent behavior.
- [ ] Checked if this PR would require a ReadMe/Wiki update.

Check which platforms you have built SDK on to verify the correctness of this PR.
- [x] Linux
- [x] Windows
- [ ] Android
- [ ] MacOS
- [ ] IOS
- [ ] Other Platforms


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

sbera87 commented:
resending

sbiscigl commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsParser.java:
```diff
@@ -0,0 +1,362 @@
+            }
+
+            //get configuration properties
+            ClientConfiguration config = new ClientConfiguration();
```
should create a static function in ClientConfiguration called `FromSmokeModelTest`. that can be called

```
ClientConfiguration::FromSmokeModelTest(AwsSmokeTestModel).getOrElseThrow(new WahteverException(...))
```

then move this logic into that static function, since config really isnt optional we can throw a exception.

sbiscigl commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppImportContainer.java:
```diff
@@ -0,0 +1,37 @@
+ import software.amazon.smithy.codegen.core.ImportContainer;
+ import software.amazon.smithy.codegen.core.Symbol;
+ 
+ public final class CppImportContainer implements ImportContainer {
```
why isnt this used to add includes, and why is it required? maybe make one for cmake and one for cpp source

sbera87 replied:
This is just a pass through , we didn't have to use this but the plugin accepts this class . So, needed to provide this

sbiscigl commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmithyC2JNamespaceMap.java:
```diff
@@ -0,0 +1,55 @@
+
+public class SmithyC2JNamespaceMap {
+    private static SmithyC2JNamespaceMap instance = null;
+    private  static Map<String, String> SMITHY_C2J_SERVICE_NAME_MAP;
```
this really isnt static because every constructor initializes it, might as well make it a member variable

sbiscigl commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppSymbolVisitor.java:
```diff
@@ -0,0 +1,194 @@
+ 
+     @Override
+     public Symbol blobShape(BlobShape blobShape) {
+         return createSymbolBuilder(blobShape, "stringstream").build();
```
need top use Aws::StringStream not std string stream

sbera87 replied:
ack

sbiscigl commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsParser.java:
```diff
@@ -0,0 +1,362 @@
+                    });
+                }
+                catch (Exception e) {
+                    System.err.println("Exception detected=" + e.toString());
```
remove catch or re-throw runtime exception so the generation will actually fail

sbera87 replied:
ack

sbiscigl commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmithyCodegenPlugin.java:
```diff
@@ -13,8 +13,13 @@ public String getName() {
 @Override
     public void execute(PluginContext context){
```
fix indentation

sbera87 replied:
ack

sbiscigl commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/GenericCodegenAdapter.java:
```diff
@@ -0,0 +1,213 @@
+
+    public String getShapeName(SHAPE s);
+
+    public static String convertSnakeToPascal(String snakeCase) {
```
static functions in a interface seems kind of sketchy. Why do they need to be in a interface? they are not being overridden so why can they not live outside of the interface? also why does this function need to exist when untility functions like it exist in [guava](https://github.com/google/guava/wiki/StringsExplained#caseformat) or [apache commons](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html). we shoudlnt be re-implementing common utilities from standard libraries

sbera87 replied:
ack. will replace

sbiscigl commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmithyCodegenAdapter.java:
```diff
@@ -0,0 +1,220 @@
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class SmithyCodegenAdapter implements GenericCodegenAdapter<Shape, Node>{
```
how does implementing `GenericCodegenAdapter` help us? i know before this in a previous iteration that it was between c2j and smithy, but it seems like a uneeded abstraction now where we could just use shape and node directly.

sbera87 replied:
So this really helps separate the implementation from the algorithm. It helps with abstraction

sbiscigl reviewed (commented):
Left some comments, have some concerns around exception handling and the structure of some of the operations. but overall the execution pattern looks good, just need to refactor some things.

also in the future make sure to remove commented code, system outs, and white space changes before marking the PR as ready to review

sbera87 commented on VERSION:
```diff
@@ -1 +1 @@
-1.11.430
```
ack. its a merge issue. should be ommited

sbera87 commented on cmake/sdks.cmake:
```diff
@@ -162,7 +162,6 @@ if(BUILD_ONLY)
         else()
             set(SDK_DIR "generated/src/aws-cpp-sdk-${SDK}")
         endif()
-
```
ack

sbera87 replied:
its very hard to see on ide :(

sbera87 commented on cmake/sdks.cmake:
```diff
@@ -214,10 +213,30 @@ function(add_sdks)
     endforeach()
 
+    file(GLOB subdirs LIST_DIRECTORIES true "${CMAKE_SOURCE_DIR}/generated/smoke-tests/*")
+    foreach(subdir ${subdirs})
```
The code honors the client filter because it looks for the client in sdk_build_list first and only then builds it

 list(FIND SDK_BUILD_LIST ${folder_name} index)

        # Check if the item was found (index >= 0)
        if(${index} GREATER -1)
            message(STATUS "${subdir} is in SDK_BUILD_LIST at index ${index}")

            if(EXISTS "${subdir}/CMakeLists.txt")
                add_subdirectory(${subdir})
            endif()
        else()
            message(STATUS "${subdir} is NOT in SDK_BUILD_LIST")
        endif()

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/ClientConfiguration.java:
```diff
@@ -0,0 +1,10 @@
+import software.amazon.smithy.aws.smoketests.model.S3VendorParams;
+@Data
+public final class ClientConfiguration {
+    Optional<AwsVendorParams> awsParams;
```
How do you model a class with optional params? Per smoke tests, there are two types of client config params: one for s3 other for rest. That's what I am trying to capture here. if i don't make optional then I have to initialize to something.

sbera87 commented on tools/scripts/smithy2c2j_service_map.json:
```diff
@@ -0,0 +1,51 @@
+{
```
There is no duplicate here after last commit

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmithyCodegenPlugin.java:
```diff
@@ -13,8 +13,13 @@ public String getName() {
+        } catch (Exception e) {
+            // Handle the exception
+            System.err.println("An error occurred while running the smoke tests: " + e.getMessage());
+            e.printStackTrace();
```
Sure. I wanted the smoke test failure to not block other build process in such a case. 
I can bubble the exception

sbiscigl replied:
the generation _should fail_. whatever invokes the generation in CI, should _not fail_. making it the callers responsibility to decide.

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsCMakeWriter.java:
```diff
@@ -0,0 +1,90 @@
+                $${PROJECT_LIBS})
+            """, awsTestSrc, awsTestSrc,folderNamespace);
+        /*
+            if (AUTORUN_UNIT_TESTS)
```
ack. i left this code in case we ever want to auto run smoke tests. I can remove it for now

sbera87 commented on cmake/sdks.cmake:
```diff
@@ -178,7 +177,7 @@ if(BUILD_ONLY)
                 set(SDK_DIR "src/aws-cpp-sdk-${SDK}")
             else()
-                set(SDK_DIR "generated/src/aws-cpp-sdk-${SDK}")
+                set(SDK_DIR "generated/src/aws-cpp-sdk-${SDK}") 
```
ack

SergeyRyabinin commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/ClientConfiguration.java:
```diff
@@ -0,0 +1,38 @@
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+package com.amazonaws.util.awsclientsmithygenerator.generators;
```
please stick to one of java code formatting styles
i.e. empty newline after package, empty newline after copyright

sbera87 replied:
ack

SergeyRyabinin commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppBlockWriter.java:
```diff
@@ -0,0 +1,208 @@
+import java.util.List;
+import lombok.Data;
+
+//Helper class for C++ code blocks for scope management
```
please elaborate

sbera87 replied:
ack

SergeyRyabinin commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppBlockWriter.java:
```diff
@@ -0,0 +1,208 @@
+public class CppBlockWriter {
+
+    public static String GetForLoopRangeInitializer(String variableName, String iterableExpr)
+    {
```
please use figure braces java style

sbera87 replied:
ack

SergeyRyabinin commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/GenericCodegenAdapter.java:
```diff
@@ -0,0 +1,217 @@
+
+            String returnType = "Aws::Vector<"+listType+">";
+            //open function body and use vector setter
+            CppBlockWriter blockWriter = new CppBlockWriter
+            (
+            CppBlockWriter.GetLambdaBlock(returnType, functionName), 0).
+            addCode(String.format("Aws::Vector<%s> %s = ",listType, varName))
+            .openCodeBlock("");
```
this code formatting makes it unread-able.

sbera87 replied:
reformatted

SergeyRyabinin commented on tools/scripts/run_code_generation.py:
```diff
@@ -456,6 +535,15 @@ def main():
+    c2j_smithy_data = {}        
+    with open(os.path.abspath(SMITHY_TO_C2J_MAP_FILE), 'r') as file:
+        smithy_c2j_data = json.load(file)
+        # Reverse the key-value pairs
+        c2j_smithy_data = {value: key for key, value in smithy_c2j_data.items()}
```
I can see you are reversing a map from the code
I'd prefer to have a comment explaining why

sbera87 replied:
ack

SergeyRyabinin commented on tools/scripts/run_code_generation.py:
```diff
@@ -456,6 +535,15 @@ def main():
     highly_refined_percent_of_cores_to_take = 0.9
     max_workers = max(1, int(highly_refined_percent_of_cores_to_take * os.cpu_count()))
+    
+    #build reverse map
+    # Open a file using 'with' and read its contents
```
I can get this information from the code.
Please put why this here in comments instead.

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/ClientConfiguration.java:
```diff
@@ -0,0 +1,38 @@
+import software.amazon.smithy.aws.smoketests.model.AwsVendorParams;
+import software.amazon.smithy.aws.smoketests.model.S3VendorParams;
+
+public final class ClientConfiguration {
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests/build.gradle.kts:
```diff
@@ -1,3 +1,9 @@
+import software.amazon.smithy.model.node.Node
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.aws.traits.ServiceTrait
+//import software.amazon.smithy.smoketests.traits.SmokeTestsTrait
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests/build.gradle.kts:
```diff
@@ -12,14 +18,98 @@ buildscript {
         classpath(codegen.aws.traits)
-        classpath(codegen.smoke.test.traits)
+        classpath(codegen.aws.smoke.test.model)
         //classpath(codegen.velocity.engine.core)
```
ack

sbera87 commented on tools/code-generation/smithy/api-descriptions/dynamodb.json:
ack, its a mistake.  was trying out custom tests in modified json. will be restored

sbera87 commented on tools/code-generation/smithy/codegen/build.gradle.kts:
```diff
@@ -13,4 +13,10 @@ allprojects {
-}
\ No newline at end of file
+}
+
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppSymbolVisitor.java:
```diff
@@ -0,0 +1,194 @@
+
+ package com.amazonaws.util.awsclientsmithygenerator.generators;
+
+ import software.amazon.smithy.codegen.core.*;
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/GenericCodegenAdapter.java:
```diff
@@ -0,0 +1,213 @@
+
+    public SHAPE getShapeFromOperation(String OperationName);
+
+    //public boolean isEnum(DATA d);
```
we actually not need this for the value.. enum value will always be string per smithy. will remove.

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/GenericCodegenAdapter.java:
```diff
@@ -0,0 +1,217 @@
+        String varName = key.toLowerCase();
+        String functionNameSuffix = convertSnakeToPascal(varName + "_lvl" + level + "_idx" + count);
+
+        //for simple types, use initializer list for narrow types
+        if ( (isFloat(value) && isFloatShape(shape)) ||
+                (isInteger(value) && isIntegerShape(shape) ) ||
+                (isBoolean(value) && isBooleanShape(shape))
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/GenericCodegenAdapter.java:
```diff
@@ -0,0 +1,217 @@
+        }
+        else
+        {
+            System.err.println(String.format("shape not supported:%s",shape));
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsParser.java:
```diff
@@ -0,0 +1,362 @@
+    final private SymbolProvider symbolProvider;
+    final private Map<ShapeId, String> operationToServiceMap;
+    final private Map<String, ServiceShape> serviceShapeMap;
+    // Static member to store the flipped map
```
old comment, will remove this

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsParser.java:
```diff
@@ -0,0 +1,362 @@
+                });
+        });
+
+        //System.out.println(String.format("Number of services with smoke tests detected=%d",serviceShapeMap.size()));
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsParser.java:
```diff
@@ -0,0 +1,362 @@
+    {   
+        if(!serviceShape.getTrait(ServiceTrait.class).isPresent())
+        {
+            //System.err.println(String.format("No service trait detected in service shape with name=%s",serviceShape.getId().getName()));
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsParser.java:
```diff
@@ -0,0 +1,362 @@
+
+    @Override
+    public void run(){
+        //System.out.println("run smoke tests parser");
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsSourceWriter.java:
```diff
@@ -0,0 +1,148 @@
+    private final String clientNamespace;
+    private final String folderNamespace;
+    private final String c2jNamespace;
+    //protected CppBlockWriter blockWriter;    
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsSourceWriter.java:
```diff
@@ -0,0 +1,148 @@
+    {
+        write("${L|}",
+        
+        "#include <aws/testing/AwsCppSdkGTestSuite.h>\n" + //
```
moved with latest commit

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsSourceWriter.java:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    /**
+        * The Factory class for creating CppWriters
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests/build.gradle.kts:
```diff
@@ -12,14 +18,98 @@ buildscript {
+                }
+            }
+
+            //println("matched filter: $sdkId")
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsParser.java:
```diff
@@ -0,0 +1,362 @@
+            //get serviceShape
+            String serviceName = operationToServiceMap.get(operationShape.getId());
+            ServiceShape serviceShape = serviceShapeMap.get(serviceName);
+            //System.out.println("OperationShape: " + operationShape.getId().getName());
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsParser.java:
```diff
@@ -0,0 +1,362 @@
+import software.amazon.smithy.aws.smoketests.model.AwsSmokeTestModel;
+
+
+//import com.amazonaws.util.awsclientsmithygenerator.generators.common.CppSymbolVisitor;
```
ack

sbiscigl commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsParser.java:
```diff
@@ -0,0 +1,362 @@
+        model.getServiceShapes().stream()
+        .filter(serviceShape -> {
+            try {
+                return serviceFilter.isEmpty() || serviceFilter.contains(toKebabCase(getServiceName(serviceShape)));
```
this try block gets removed if you remove the other explicitly throw `Exceptions`

sbiscigl replied:
nope you don't need to catch runtime exceptions in lambda, thats one of the best parts of them and why we love them so we can avoid that, you can remove `throws RuntimeException` from `getServiceName` and you should be able to remove all of this.

throw runtime, dont mark as throws on the funciton, let it bubble up.

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsParser.java:
```diff
@@ -0,0 +1,351 @@
+    {   
+        if(!serviceShape.getTrait(ServiceTrait.class).isPresent())
+        {
+            throw new Exception(String.format("No service trait detected in service shape with name=%s",serviceShape.getId().getName()));
```
What is your recommendation how I should handle this? I understand that stack unwinding will happen on exception but  what do i return when service trait is missing

SergeyRyabinin replied:
they have a concept of "checked and unchecked exceptions" in java.
additionally, simple regular `Exception` is too broad and the exception handling error code must catch it and do it's logic accordingly to the exception string message (vs simply type of the exception).

sbera87 replied:
ack. I will switch to runtime exception as it is one of those cases where I have to throw if service trait not found

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsParser.java:
```diff
@@ -0,0 +1,362 @@
+            throw new Exception(String.format("No service trait detected in service shape with name=%s",serviceShape.getId().getName()));
+        }
+
+        //this.model.getShape(serviceShapeId)
```
ack. Oversight

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppImportContainer.java:
```diff
@@ -0,0 +1,108 @@
+        this.coreHeaders = new HashSet<>();
+        this.dynamicHeaders = new HashSet<>();
+        Collections.addAll(coreHeaders, 
+        "aws/core/client/AsyncCallerContext.h",
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/ClientConfiguration.java:
```diff
@@ -0,0 +1,38 @@
+import software.amazon.smithy.aws.smoketests.model.S3VendorParams;
+
+public final class ClientConfiguration {
+    private final AwsVendorParams awsParams;
```
ack. changed to base class and use instance of.

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/GenericCodegenAdapter.java:
```diff
@@ -0,0 +1,201 @@
+    {
+        if(shape == null)
+        {
+            throw new Exception("Invalid shape found");
```
ack

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppImportContainer.java:
```diff
@@ -0,0 +1,108 @@
+        "algorithm"
+        );
+
+        this.unitTestHeaders = new HashSet<>();
```
In this case, it's a set .So I will just use Set.of()

sbera87 commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppSymbolVisitor.java:
```diff
@@ -0,0 +1,219 @@
+     @Override
+     public Symbol enumShape(EnumShape shape) {
+        
+        // return ShapeVisitor.super.enumShape(shape);
```
ack. oversight

sbera87 commented:
> Left some comments, have some concerns around exception handling and the structure of some of the operations. but overall the execution pattern looks good, just need to refactor some things.
> 
> also in the future make sure to remove commented code, system outs, and white space changes before marking the PR as ready to review

Addressed all the comments that were brought up. If there are some things missed, I will address

sbiscigl commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsParser.java:
```diff
@@ -0,0 +1,348 @@
+                                        functionOrder)
+                                ));
+                        } catch (RuntimeException e) {
+                            throw new RuntimeException(e);
```
remove the try catch block completely since we dont need to catch to re-throw

sbera87 replied:
ok, that will mean first exception will cause exit. If that's indeed the idea, I will remove it

sbiscigl replied:
yeah thats the idea, if we throw a exception we want to fail codegen, let the caller of codegen decide what to do after we know it has failed

sbiscigl commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsParser.java:
```diff
@@ -0,0 +1,348 @@
+                delegator.flushWriters();
+                cmakedelegator.flushWriters();
+            } catch (Exception e) {
+                System.err.println("Exception detected=" + e.toString());
```
dont catch the exception, let the exception bubble up as a runtime, and just let it fail if a exception is thrown

sbera87 replied:
ok, removed. The only thing is that first exception will bubble up to exit

sbiscigl replied:
thats fine thats what we want

sbiscigl commented on tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestData.java:
```diff
@@ -0,0 +1,26 @@
+    public boolean expectSuccess;
+    public Optional<String> errorShapeId;
+    //capture auth scheme as that decides the client constructor
+    String auth;
```
nit: why is this member the only one with a comment and the only one without access modifier?

sbera87 replied:
ack. missed the specifier, shall add . The comment for this field is because this is a derived field, not parsed directly from the smoke test trait in model since its available in the service trait
